skimage/morphology/extrema.py

```diff
@@ -207,152 +210,360 @@ def h_minima(image, h, selem=None):
-    4 additional constant maxima.
-    The heights of the maxima are: 1, 21, 41, 61, 81, 101
+    Notes
+    -----
+    This function operates on the following ideas:
 
-    >>> w = 10
-    >>> x, y = np.mgrid[0:w,0:w]
-    >>> f = 20 - 0.2*((x - w/2)**2 + (y-w/2)**2)
-    >>> f[2:4,2:4] = 40; f[2:4,7:9] = 60; f[7:9,2:4] = 80; f[7:9,7:9] = 100
-    >>> f = f.astype(np.int)
+    1. Make a first pass over the image's last dimension and flag candidates
+       for local maxima by comparing pixels in only one direction.
 
-    We can calculate all local maxima:
+    For each candidate:
 
-    >>> maxima = extrema.local_maxima(f)
+    2. Perform a flood-fill to find all connected pixels that have the same
+       gray value and are part of the plateau.
+    3. Consider the connected neighborhood of a plateau: if any no bordering
+       sample has a higher gray level, mark the plateau as a definite local
+       maximum.
 
-    The resulting image will contain all 6 local maxima.
+    Examples
+    --------
+    >>> from skimage.morphology import local_maxima
+    >>> image = np.zeros((4, 7), dtype=int)
+    >>> image[1:3, 1:3] = 1
+    >>> image[3, 0] = 1
+    >>> image[1:3, 4:6] = 2
+    >>> image[3, 6] = 3
+    >>> image
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 1, 1, 0, 2, 2, 0],
+           [0, 1, 1, 0, 2, 2, 0],
+           [1, 0, 0, 0, 0, 0, 3]])
+
+    Find local maxima by comparing to all neighboring pixels (maximal
+    connectivity):
+
+    >>> local_maxima(image)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 1, 1, 0, 0, 0, 0],
+           [0, 1, 1, 0, 0, 0, 0],
+           [1, 0, 0, 0, 0, 0, 1]], dtype=uint8)
+    >>> local_maxima(image, indices=True)
+    (array([1, 1, 2, 2, 3, 3]), array([1, 2, 1, 2, 0, 6]))
+
+    Find local maxima without comparing to diagonal pixels (connectivity 1):
+
+    >>> local_maxima(image, selem=1)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 1, 1, 0, 1, 1, 0],
+           [0, 1, 1, 0, 1, 1, 0],
+           [1, 0, 0, 0, 0, 0, 1]], dtype=uint8)
+
+    and exclude maxima that border the image edge:
+
+    >>> local_maxima(image, selem=1, include_border=False)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 1, 1, 0, 1, 1, 0],
+           [0, 1, 1, 0, 1, 1, 0],
+           [0, 0, 0, 0, 0, 0, 0]], dtype=uint8)
     """
-    # find the minimal grey level difference
-    h = _find_min_diff(image)
-    if h == 0:
-        return np.zeros(image.shape, np.uint8)
-    if not np.issubdtype(image.dtype, np.floating):
-        h = 1
-    local_max = h_maxima(image, h, selem=selem)
-    return local_max
+    image = np.asarray(image)
+    if image.size == 0:
+        # Return early for empty input
+        return np.array([], dtype=(np.intp if indices else np.uint8))
+
+    if include_border:
+        # Ensure that local maxima are always at least one smaller sample away
+        # from the image border
+        image = _fast_pad(image, image.min())
+
+    if selem is None:
```

Comment:
Can you abstract all this logic into a private function? It's quite cluttery. Something like `selem = _validate_selem(image, selem, connectivity)`